myokit/formats/diffsl/_exporter.py

```diff
@@ -0,0 +1,523 @@
+                'Not a variable or LhsExpression: ' + str(e)
+            )
+
+        # Create an expression writer
+        e = diffsl.DiffSLExpressionWriter()
+        e.set_lhs_function(var_name)
+
+        export_lines = []  # DiffSL export lines
+        tab = '  '  # Tab character
+
+        # Sort equations in solvable order (grouped by component)
+        sorted_eqs = model.solvable_order()
+
+        # Variables to be excluded from output or handled separately.
+        # Derivatives and their intermediary variables (i.e. dot(x) and dot_x)
+        # are handled in dudt_i, F_i and G_i; time is excluded from the output;
+        # pace is made Vc.
+        time = model.time()
+        pace = model.binding('pace')
+        special_vars = set(
+            [time]
+            + [v.rhs().var() for v in model.states()]
+            + [v.lhs().var() for v in model.states()]
+        )
+        if pace is not None:
+            special_vars.add(pace)
+
+        # Add metadata
+        export_lines.append('/*')
+        export_lines.append('This file was generated by Myokit.')
+        if model.meta:
+            export_lines.append('')
+            for key, val in sorted(model.meta.items()):
+                export_lines.append(f'{key}: {val}')
+            export_lines.append('')
+        export_lines.append('*/')
+        export_lines.append('')
+
+        # Add empty input parameter list
+        export_lines.append('/* Input parameters */')
+        export_lines.append('in = [ ]')
+        export_lines.append('')
+
+        # Add placeholder protocol
+        if pace is not None:
+            export_lines.append('/* Voltage protocol [mV] */')
+            export_lines.append(
+                'Vc { -80 + 120 * heaviside(t-500) - 80 * heaviside(t-1000) }'
+            )
+            export_lines.append('')
+
+        # Add constants
+        const_vars = (
+            set(model.variables(const=True, deep=True, state=False))
+            - special_vars
+        )
+        for component_label, eq_list in sorted_eqs.items():
+            const_eqs = [
+                eq for eq in eq_list.equations() if eq.lhs.var() in const_vars
+            ]
+            if const_eqs:
+                export_lines.append(f'/* Constants: {component_label} */')
+                for eq in const_eqs:
+                    v = eq.lhs.var()
+                    lhs = var_name(v)
+                    rhs = e.ex(eq.rhs)
+                    qname = v.qname()
+                    unit = '' if v.unit() is None else f' {v.unit()}'
+                    export_lines.append(
+                        f'{lhs} {{ {rhs} }} /* {qname}{unit} */'
+                    )
+                export_lines.append('')
+
+        # Add initial conditions `u_i`
+        export_lines.append('/* Initial conditions */')
+        export_lines.append('u_i {')
+        for v in model.states():
+            lhs = var_name(v)
+            rhs = myokit.float.str(v.initial_value(as_float=True))
```

Comment:
Does DiffSL support expressions in initial values? It was martin who implemented that feature in Myokit, so might want to maintain expressions here instead of converting to floats?

Reply:
(Note these expressions can only depend on constants)

Reply:
Resolved by 1a21249